File: AircraftWar/src/main/scala/work/chiro/game/application/Game.scala
package work.chiro.game.application

import work.chiro.game.GlobalConfigLoader.config
import work.chiro.game.aircraft._
import work.chiro.game.basic.AbstractObject
import work.chiro.game.bullet.AbstractBullet
import work.chiro.game.control.HeroController
import work.chiro.game.logger
import work.chiro.game.prop.{AbstractProp, BloodProp, BombProp, BulletProp}
import work.chiro.game.scene.Background
import work.chiro.game.utils.getTimeMills

import java.awt.geom.AffineTransform
import java.awt.{Color, Font, Graphics, Graphics2D}
import java.util.concurrent.{ScheduledThreadPoolExecutor, TimeUnit}
import javax.swing._
import scala.collection.mutable.ListBuffer

/**
 * 游戏主面板，游戏启动
 *
 * @author chiro2001
 */
class Game(frame: JFrame) extends JPanel {
  logger.info(s"Window(${config.window.width}x${config.window.height})")
  val heroAircraft = HeroAircraft.create()
  val heroPosition = HeroAircraft.getPositionInstance
  val enemyAircrafts = new ListBuffer[AbstractAircraft]
  val heroBullets = new ListBuffer[AbstractBullet]
  val enemyBullets = new ListBuffer[AbstractBullet]
  val props = new ListBuffer[AbstractProp]
  val allObjectLists = Array(
    enemyAircrafts,
    heroBullets,
    enemyBullets,
    props
  )
  val allAircrafts = Array(enemyAircrafts, List(heroAircraft))
  val allBullets = Array(heroBullets, enemyBullets)
  // Scheduled 线程池，用于定时任务调度
  val executorService = new ScheduledThreadPoolExecutor(1)
  // 启动英雄机控制监听
  val controller = new HeroController(frame, this, heroAircraft)
  private var backGroundTop = 0
  private val timeInterval = 1
  private var gameOverFlag = false
  private var score = 0

  private var frameTime: Double = 0
  private var lastFrameTime: Double = 0

  private val heroShootDuration = 1
  private var heroShootCycleTime: Double = 0

  private val mobCreateDuration = 1200
  private var mobCreateCycleTime: Double = 0

  private val eliteCreateDuration = 600
  private var eliteCreateCycleTime: Double = 0

  private val eliteShootDuration = 200
  private var eliteShootCycleTime: Double = 0

  private var fpsCycleTime: Double = 0
  private val fpsCycleDuration = 100
  private val frameCount = new ListBuffer[Double]

  def frameTimeDelta = frameTime - lastFrameTime

  /**
   * 游戏启动入口，执行游戏逻辑
   */
  def action() = {
    enemyAircrafts.synchronized(enemyAircrafts.append(EliteEnemy.create()))
    // 定时任务：绘制、对象产生、碰撞判定、击毁及结束判定
    val calcTask = new Runnable {
      override def run() = {
        // 周期性执行（控制频率）
        // 键盘操作
        controller.onFrame()
        // 英雄飞机射出子弹
        if (onHeroShootCountCycle) heroShootAction()
        // 敌机射出子弹
        if (onEliteShootCountCycle) eliteShootAction()
        // 新敌机产生
        if (onMobCreateCountCycle) enemyAircrafts.synchronized(enemyAircrafts.append(MobEnemy.create()))
        // 产生精英敌机
        if (onEliteCreateCountCycle) enemyAircrafts.synchronized(enemyAircrafts.append(EliteEnemy.create()))
        if (onFpsCountCycle) {
          frameCount.synchronized({
            val fpsInfo = f"[ ${frameTime / 1000}%.3fs ] ${frameCount.size} fps"
            if (config.running.showFps) logger.info(fpsInfo)
            Main.getFrameInstance.get.setTitle(f"Aircraft War $fpsInfo")
          })
        }
        // 所有物体移动
        allObjectLists.foreach(_.foreach(_.forward()))
        // 撞击检测
        crashCheckAction()
        // 后处理
        postProcessAction()
        // 游戏结束检查
        if (!config.isDebug && heroAircraft.getHp <= 0) { // 游戏结束
          executorService.shutdown()
          gameOverFlag = true
          logger.info("Game Over!")
        }
        lastFrameTime = frameTime
      }
    }

    val renderTask = new Runnable {
      override def run() = {
        frameTime = getTimeMills
        frameCount.synchronized({
          frameCount.append(frameTime)
          frameCount.filterInPlace(_ >= (if (frameTime >= 1000) frameTime - 1000 else 0))
        })
        // 重绘界面
        repaint()
      }
    }

    /**
     * 以固定延迟时间进行执行
     * 本次任务执行完成后，需要延迟设定的延迟时间，才会执行新的任务
     */
    executorService.scheduleWithFixedDelay(calcTask, timeInterval, timeInterval, TimeUnit.MILLISECONDS)
    executorService.scheduleWithFixedDelay(renderTask, timeInterval, timeInterval, TimeUnit.MILLISECONDS)
    // 不管上次执行是否完成也执行
    // executorService.scheduleAtFixedRate(calcTask, timeInterval, timeInterval, TimeUnit.MILLISECONDS)
  }

  private def onHeroShootCountCycle = {
    heroShootCycleTime += frameTimeDelta
    // 跨越到新的周期
    if (heroShootCycleTime >= heroShootDuration) {
      heroShootCycleTime %= heroShootDuration
      true
    } else false
  }

  private def onEliteShootCountCycle = {
    eliteShootCycleTime += frameTimeDelta
    // 跨越到新的周期
    if (eliteShootCycleTime >= eliteShootDuration) {
      eliteShootCycleTime %= eliteShootDuration
      true
    } else false
  }

  private def onMobCreateCountCycle = {
    mobCreateCycleTime += frameTimeDelta
    // 跨越到新的周期
    if (mobCreateCycleTime >= mobCreateDuration) {
      mobCreateCycleTime %= mobCreateDuration
      true
    } else false
  }

  private def onEliteCreateCountCycle = {
    eliteCreateCycleTime += frameTimeDelta
    // 跨越到新的周期
    if (eliteCreateCycleTime >= eliteCreateDuration) {
      eliteCreateCycleTime %= eliteCreateDuration
      true
    } else false
  }

  private def onFpsCountCycle = {
    fpsCycleTime += frameTimeDelta
    // 跨越到新的周期
    if (fpsCycleTime >= fpsCycleDuration) {
      // logger.info(f"bullet count: ${allBullets.map(_.size).sum}")
      fpsCycleTime %= fpsCycleDuration
      true
    } else false
  }

  private def heroShootAction() = {
    // 英雄射击
    if (controller.pressed(config.control.keyShoot))
      heroBullets.synchronized(heroBullets.addAll(heroAircraft.shoot()))
  }

  private def eliteShootAction() = {
    // 精英敌机射击
    enemyAircrafts.foreach(enemy => enemyBullets.synchronized(enemyBullets.addAll(enemy.shoot())))
  }

  /**
   * 碰撞检测：
   * 1. 敌机攻击英雄
   * 2. 英雄攻击/撞击敌机
   * 3. 英雄获得补给
   */
  private def crashCheckAction() = {
    // 敌机子弹攻击英雄
    enemyBullets.foreach(bullet => if (bullet.isValid) {
      if (heroAircraft.crash(bullet)) {
        heroAircraft.decreaseHp(bullet.getPower)
        bullet.vanish()
      }
    })
    // 英雄子弹攻击敌机
    heroBullets.foreach(bullet => {
      if (bullet.isValid)
        enemyAircrafts.foreach(enemyAircraft => {
          if (enemyAircraft.isValid) { // 已被其他子弹击毁的敌机，不再检测
            // 避免多个子弹重复击毁同一敌机的判定
            if (enemyAircraft.crash(bullet)) { // 敌机撞击到英雄机子弹
              // 敌机损失一定生命值
              enemyAircraft.decreaseHp(bullet.getPower)
              bullet.vanish()
              if (!enemyAircraft.isValid) {
                // 获得分数
                score += 10
                if (enemyAircraft.getClass.getName.endsWith("EliteEnemy")) {
                  // 产生道具补给
                  if (math.random() < 0.7) {
                    props.synchronized(props.append((math.random() * 3).toInt match {
                      case 0 => BulletProp.create(enemyAircraft.getPos)
                      case 1 => BombProp.create(enemyAircraft.getPos)
                      case 2 => BloodProp.create(enemyAircraft.getPos)
                    }))
                  }
                }
              }
            }
            // 英雄机 与 敌机 相撞，均损毁
            if (enemyAircraft.crash(heroAircraft) || heroAircraft.crash(enemyAircraft)) {
              enemyAircraft.vanish()
              heroAircraft.decreaseHp(Integer.MAX_VALUE)
            }
          }
        })
    })
    // 我方获得道具
    props.foreach(prop => if (prop.isValid) {
      if (prop.crash(heroAircraft, useImageBox = true)) {
        // 道具生效
        prop.handleAircrafts(enemyAircrafts.toList)
        prop.vanish()
      }
    })
  }

  /**
   * 后处理：
   * 1. 删除无效的子弹
   * 2. 删除无效的敌机
   * 3. 检查英雄机生存
   * 无效的原因可能是撞击或者飞出边界
   */
  private def postProcessAction() = {
    allObjectLists.foreach(objList => objList.synchronized(objList.filterInPlace(_.isValid)))
  }

  /**
   * 重写paint方法
   * 通过重复调用paint方法，实现游戏动画
   */
  override def paint(g: Graphics) = {
    super.paint(g)
    // 绘制背景,图片滚动
    g.drawImage(Background.getImage, 0, backGroundTop - config.window.height, null)
    g.drawImage(Background.getImage, 0, backGroundTop, null)
    // backGroundTop += 1
    if (backGroundTop == config.window.height) backGroundTop = 0
    // 先绘制子弹，后绘制飞机
    // 这样子弹显示在飞机的下层
    paintImageWithPositionRevised(g, heroBullets)
    paintImageWithPositionRevised(g, enemyAircrafts)
    g.drawImage(
      HeroAircraft.getImage,
      (heroAircraft.getLocationX - HeroAircraft.getImage.getWidth / 2).toInt,
      (heroAircraft.getLocationY - HeroAircraft.getImage.getHeight / 2).toInt,
      null
    )
    paintImageWithPositionRevised(g, enemyBullets)
    paintImageWithPositionRevised(g, props)
    g.drawImage(
      HeroAircraft.HeroBox.getImage,
      (heroAircraft.getLocationX - config.hero.box / 2).toInt,
      (heroAircraft.getLocationY - config.hero.box / 2).toInt,
      config.hero.box,
      config.hero.box,
      null
    )
    //绘制得分和生命值
    paintScoreAndLife(g)
  }

  private def paintImageWithPositionRevised(g: Graphics, objects: ListBuffer[_ <: AbstractObject]): Unit = {
    objects.synchronized {
      objects.foreach(obj => {
        val image = obj.getImage
        if (obj.rotation.getX == 0) {
          g.drawImage(image, (obj.getLocationX - image.getWidth / 2).toInt, (obj.getLocationY - image.getHeight / 2).toInt, null)
        } else {
          val af = AffineTransform.getTranslateInstance(obj.getLocationX - image.getWidth / 2, obj.getLocationY - image.getHeight / 2)
          af.rotate(obj.rotation.getX, image.getWidth / 2, image.getHeight / 2)
          val graphics2d = g.asInstanceOf[Graphics2D]
          graphics2d.drawImage(image, af, null)
        }
      })
    }
  }

  private def paintScoreAndLife(g: Graphics) = {
    val x = 10
    var y = 25
    g.setColor(new Color(16711680))
    g.setFont(new Font("SansSerif", Font.BOLD, 22))
    g.drawString("SCORE:" + this.score, x, y)
    y = y + 20
    g.drawString("LIFE:" + this.heroAircraft.getHp, x, y)
  }
}

File: app/src/main/java/work/chiro/game/game/Game.java
package work.chiro.game.game;

import work.chiro.game.aircraft.*;
import work.chiro.game.application.HeroController;
import work.chiro.game.application.MusicManager;
import work.chiro.game.application.MusicThreadFactory;
import work.chiro.game.application.MyThreadFactory;
import work.chiro.game.background.*;
import work.chiro.game.basic.AbstractFlyingObject;
import work.chiro.game.bullet.BaseBullet;
import work.chiro.game.config.*;
import work.chiro.game.dao.HistoryImpl;
import work.chiro.game.dao.HistoryObjectFactory;
import work.chiro.game.prop.AbstractProp;
import work.chiro.game.scene.SceneRun;
import work.chiro.game.timer.Timer;
import work.chiro.game.timer.TimerController;
import work.chiro.game.utils.Utils;
import work.chiro.game.windows.HistoryWindow;

import javax.swing.*;
import java.awt.*;
import java.io.IOException;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

/**
 * 游戏主面板，游戏启动
 *
 * @author hitsz
 */
public class Game extends JPanel {
    /**
     * 创建线程的工厂函数
     */
    static private final MyThreadFactory THREAD_FACTORY = new MyThreadFactory("AircraftWar");
    static private final MusicThreadFactory MUSIC_FACTORY = new MusicThreadFactory("AircraftWar-Music");
    /**
     * 线程池，自动管理
     */
    @SuppressWarnings("AlibabaThreadPoolCreation")
    static private final ScheduledExecutorService executorService = Executors.newScheduledThreadPool(Constants.GAME_POOL_SIZE, THREAD_FACTORY);

    public static MyThreadFactory getThreadFactory() {
        return THREAD_FACTORY;
    }

    public static MusicThreadFactory getMusicFactory() {
        return MUSIC_FACTORY;
    }

    private HeroAircraft heroAircraft;
    private final List<AbstractAircraft> heroAircrafts = new LinkedList<>();
    private final List<AbstractBackground> backgrounds = new LinkedList<>();
    private final List<BossEnemy> bossAircrafts = new LinkedList<>();
    private final List<AbstractAircraft> enemyAircrafts = new LinkedList<>();
    private final List<BaseBullet> heroBullets = new LinkedList<>();
    private final List<BaseBullet> enemyBullets = new LinkedList<>();
    private final List<AbstractProp> props = new LinkedList<>();
    private final List<List<? extends AbstractFlyingObject>> allObjects = Arrays.asList(
            backgrounds, heroBullets, heroAircrafts, enemyBullets, enemyAircrafts, bossAircrafts, props
    );
    private boolean gameOverFlag = false;
    private boolean startedFlag = false;
    private double score = 0;
    private double nextBossScore;
    private final Object waitObject = new Object();
    private Future<?> future = null;
    private Font myFontBase = null;
    private String lastProvidedName = null;
    private String lastProvidedMessage = null;
    private final HeroController heroController;
    private final TimerController timerController = new TimerController();
    private AbstractConfig config;
    private Difficulty difficulty;

    public void resetStates() {
        gameOverFlag = false;
        score = 0;
        enemyBullets.clear();
        enemyAircrafts.clear();
        props.clear();
        difficulty = RunningConfig.difficulty;
        config = new ConfigFactory(difficulty).create();
        heroAircraft = new HeroAircraftFactory().clearInstance().create(config);
        heroAircrafts.clear();
        heroAircrafts.add(heroAircraft);
        bossAircrafts.clear();
        BossEnemyFactory.clearInstance();
        timerController.clear();
        heroController.clear();

        flushBackground();
    }

    public boolean getGameOverFlag() {
        return gameOverFlag;
    }

    public boolean getStartedFlag() {
        return startedFlag;
    }

    public Object getWaitObject() {
        return waitObject;
    }

    /**
     * 周期（ms)
     * 指示子弹的发射、敌机的产生频率
     */
    public Game(Difficulty difficulty) {
        this.difficulty = difficulty;
        config = new ConfigFactory(difficulty).create();
        nextBossScore = score + config.getBossScoreThreshold().getScaleNow().getX();
        heroAircraft = new HeroAircraftFactory().create(config);
        heroAircrafts.add(heroAircraft);
        loadFont();
        flushBackground();
        // 启动英雄机鼠标监听
        heroController = HeroController.getInstance(this);
        System.out.println("Game instance created!");
    }

    private void flushBackground() {
        backgrounds.clear();
        switch (difficulty) {
            case Easy:
                backgrounds.add(new OtherBackgroundFactory<>(new EasyBackground()).create());
                break;
            case Medium:
                backgrounds.add(new OtherBackgroundFactory<>(new MediumBackground()).create());
                break;
            case Hard:
                backgrounds.add(new OtherBackgroundFactory<>(new HardBackground()).create());
                break;
            default:
                break;
        }
    }

    private void heroShoot() {
        synchronized (heroBullets) {
            heroBullets.addAll(heroAircraft.shoot(List.of(enemyAircrafts, bossAircrafts)));
        }
    }

    public void addEvents() {
        MusicManager.initAll();
        timerController.init(Utils.getTimeMills());
        // 英雄射击事件
        timerController.add(new Timer(config.getHeroShoot(), () -> {
            if (RunningConfig.autoShoot) {
                heroShoot();
            } else {
                if (HeroController.getInstance(this).pressed(HeroController.KeyCode.SHOOT)) {
                    heroShoot();
                }
            }
        }));
        // 产生精英敌机事件
        timerController.add(new Timer(config.getEliteCreate(), () -> {
            synchronized (enemyAircrafts) {
                enemyAircrafts.add(new EliteEnemyFactory().create(config));
            }
        }));
        // 产生普通敌机事件
        timerController.add(new Timer(config.getMobCreate(), () -> {
            synchronized (enemyAircrafts) {
                enemyAircrafts.add(new MobEnemyFactory().create(config));
            }
        }));
        // 敌机射击事件
        timerController.add(new Timer(config.getEnemyShoot(), () -> {
            synchronized (enemyBullets) {
                enemyAircrafts.forEach(enemyAircraft -> enemyBullets.addAll(enemyAircraft.shoot()));
            }
        }));
        // boss射击事件
        timerController.add(new Timer(config.getBossShoot(), () -> {
            synchronized (enemyBullets) {
                bossAircrafts.forEach(bossEnemy -> enemyBullets.addAll(bossEnemy.shoot()));
            }
        }));
        // boss 生成事件
        timerController.add(new Timer(10, () -> {
            config.getBossScoreThreshold().update(Utils.getTimeMills());
            if (score > nextBossScore && bossAircrafts.isEmpty()) {
                synchronized (bossAircrafts) {
                    bossAircrafts.add(new BossEnemyFactory(() -> {
                        nextBossScore = config.getBossScoreThreshold().getScaleNow().getX() + score;
                        BossEnemyFactory.clearInstance();
                    }).create(config));
                }
            }
        }));
        // 获取键盘焦点
        timerController.add(new Timer(100, this::requestFocus));
        // 输出当前 config
        if (RunningConfig.difficulty != Difficulty.Easy) {
            timerController.add(new Timer(2000, () -> config.printNow()));
        } else {
            System.out.println("简单模式 Config 将不会改变: " + config);
        }
    }

    private void stopAllMusic() {
        System.out.println("stopping all music");
        getMusicFactory().interruptAll();
    }

    private void onGameOver() {
        // 游戏结束
        if (future != null) {
            future.cancel(true);
        }
        // 游戏结束
        gameOverFlag = true;
        Utils.startMusic(MusicManager.MusicType.GAME_OVER, true);
        stopAllMusic();
        System.out.println("Game Over!");
        try {
            String name = JOptionPane.showInputDialog("输入你的名字", lastProvidedName == null ? "Nanshi" : lastProvidedName);
            if (name == null) {
                String name2 = JOptionPane.showInputDialog("输入你的名字", lastProvidedName == null ? "Nanshi" : lastProvidedName);
                if (name2 == null) {
                    int res = JOptionPane.showConfirmDialog(null, "不保存记录?", "Save Game", JOptionPane.OK_CANCEL_OPTION);
                    if (res == JOptionPane.YES_OPTION) {
                        throw new Exception();
                    }
                } else {
                    lastProvidedName = name2;
                }
            } else {
                lastProvidedName = name;
            }
            String message = JOptionPane.showInputDialog("输入额外的信息", lastProvidedMessage == null ? "NO MESSAGE" : lastProvidedMessage);
            lastProvidedMessage = message;
            // 保存游戏结果
            if (score > 0) {
                HistoryImpl.getInstance().addOne(
                        new HistoryObjectFactory(
                                name == null ? "Nanshi" : name.isEmpty() ? "Nanshi" : name,
                                score,
                                message == null ? "NO MESSAGE" : message.isEmpty() ? "NO MESSAGE" : message,
                                difficulty)
                                .create());
            }
        } catch (Exception e) {
            e.printStackTrace();
            System.out.println("Input exception: " + e);
        } finally {
            SceneRun.getInstance().setNextScene(HistoryWindow.class);
            synchronized (waitObject) {
                waitObject.notify();
            }
        }
    }

    protected Runnable getMainTask() {
        return () -> {
            try {
                timerController.update();
                // execute all
                timerController.execute();
                // 按键处理
                heroController.onFrame();
                // 所有物体移动
                synchronized (allObjects) {
                    allObjects.forEach(objList -> objList.forEach(AbstractFlyingObject::forward));
                }
                // 撞击检测
                crashCheckAction();
                // 后处理
                synchronized (allObjects) {
                    allObjects.forEach(objList -> objList.removeIf(AbstractFlyingObject::notValid));
                }
                // 每个时刻重绘界面
                repaint();
                // 游戏结束检查和处理
                if (heroAircraft.getHp() <= 0 && !Constants.DEBUG_NO_DEATH) {
                    onGameOver();
                }
                timerController.done();
                Thread.sleep(1);
            } catch (InterruptedException e) {
                System.out.println("this thread will exit: " + e);
            }
        };
    }

    /**
     * 游戏启动入口，执行游戏逻辑
     */
    public void action() {
        startedFlag = true;
        addEvents();
        Utils.startLoopMusic(MusicManager.MusicType.BGM);
        // 定时任务：绘制、对象产生、碰撞判定、击毁及结束判定
        Runnable mainTask = getMainTask();
        int timeInterval = 1;
        // 以固定延迟时间进行执行本次任务执行完成后，需要延迟设定的延迟时间，才会执行新的任务
        future = executorService.scheduleWithFixedDelay(mainTask, timeInterval, timeInterval, TimeUnit.MILLISECONDS);
    }

    private void onEnemyAircraftHit(AbstractAircraft enemyAircraft, BaseBullet bullet) {
        enemyAircraft.decreaseHp(bullet.getPower());
        bullet.vanish();
        if (enemyAircraft.notValid()) {
            score += enemyAircraft.getScore();
            props.addAll(enemyAircraft.dropProps().stream().map(
                    prop -> prop.subscribeEnemyAircrafts(enemyAircrafts)
                            .subscribeEnemyBullets(enemyBullets)
            ).collect(Collectors.toList()));
        }
    }

    /**
     * 碰撞检测：
     * 1. 敌机攻击英雄
     * 2. 英雄攻击/撞击敌机
     * 3. 英雄获得补给
     */
    private void crashCheckAction() {
        BossEnemy boss = BossEnemyFactory.getInstance();
        // 英雄子弹攻击敌机
        for (BaseBullet bullet : heroBullets) {
            if (bullet.notValid()) {
                continue;
            }
            if (boss != null && !boss.notValid()) {
                if (bullet.crash(boss)) {
                    onEnemyAircraftHit(boss, bullet);
                    continue;
                }
            }
            for (AbstractAircraft enemyAircraft : enemyAircrafts) {
                if (enemyAircraft.notValid()) {
                    // 已被其他子弹击毁的敌机，不再检测
                    // 避免多个子弹重复击毁同一敌机的判定
                    continue;
                }
                if (enemyAircraft.crash(bullet)) {
                    // 敌机撞击到英雄机子弹
                    // 敌机损失一定生命值
                    onEnemyAircraftHit(enemyAircraft, bullet);
                }
                // 英雄机 与 敌机 相撞，均损毁
                if (heroAircraft.crash(enemyAircraft)) {
                    enemyAircraft.vanish();
                    heroAircraft.decreaseHp(Integer.MAX_VALUE);
                }
            }
        }

        // 敌机子弹攻击英雄
        for (BaseBullet bullet : enemyBullets) {
            if (bullet.notValid()) {
                continue;
            }
            if (heroAircraft.crash(bullet)) {
                heroAircraft.decreaseHp(bullet.getPower());
                bullet.vanish();
            }
        }

        // 我方获得道具，道具生效
        for (AbstractProp prop : props) {
            if (prop.notValid()) {
                continue;
            }
            if (prop.crash(heroAircraft)) {
                prop.update();
                prop.vanish();
            }
        }

        // 英雄与 Boss 相撞
        if (boss != null && !boss.notValid() && heroAircraft.crash(boss)) {
            heroAircraft.decreaseHp(Integer.MAX_VALUE);
        }
    }

    /**
     * 重写paint方法
     * 通过重复调用paint方法，实现游戏动画
     *
     * @param g 绘图
     */
    @Override
    public void paint(Graphics g) {
        super.paint(g);

        // 绘制所有物体
        synchronized (allObjects) {
            allObjects.forEach(objList -> {
                //noinspection SynchronizationOnLocalVariableOrMethodParameter
                synchronized (objList) {
                    objList.forEach(obj -> obj.draw(g));
                }
            });
        }

        //绘制得分和生命值
        paintInfo(g);

    }

    private void paintInfo(Graphics g) {
        int x = 10;
        int y = 25;
        g.setColor(new Color(0xcfcfcfcf));
        g.setFont(myFontBase);
        g.drawString("SCORE:" + (int) (this.score), x, y);
        y = y + 20;
        g.drawString("LIFE:" + (int) (this.heroAircraft.getHp()), x, y);
        y = y + 20;
        BossEnemy boss = BossEnemyFactory.getInstance();
        if (boss == null) {
            g.drawString("Before Boss:" + (int) (nextBossScore - score), x, y);
        } else {
            g.drawString("BOSS LIFE:" + (int) (boss.getHp()), x, y);
        }
        y = y + 20;
        g.drawString("FPS:" + timerController.getFps(), x, y);
    }

    private void loadFont() {
        try {
            myFontBase = Font.createFont(Font.TRUETYPE_FONT, Objects.requireNonNull(getClass().getResourceAsStream("/fonts/Genshin.ttf"))).deriveFont(22f);
        } catch (FontFormatException | IOException e) {
            myFontBase = new Font("SansSerif", Font.PLAIN, 22);
        }
    }

    public void increaseScore(double increase) {
        score += increase;
    }
}
